feat: Integrate database connectors for WBR data sourcing

OElesin opened the pull request:
This commit introduces the capability for the WBR application to fetch metrics data directly from databases, supporting Amazon Athena, PostgreSQL, Redshift, and Snowflake.

Key changes:
- Added a `connections.yaml` file to define database connection parameters.
- Modified the main WBR YAML configuration to include a `data_sources` section, where users specify a connection name (from `connections.yaml`), a SQL query, and the date column from the query result.
- Implemented database connector classes in `src/connectors/` for each supported database, along with a base connector and a factory function.
- Refactored `WBRValidator` to use these connectors to load data into a pandas DataFrame, replacing the previous CSV file input mechanism.
- Updated `controller.py` to handle the new configuration flow, loading `connections.yaml` at startup and passing data source definitions to the validator.
- Added unit tests for the new connectors, configuration loading, and related utility functions.
- Updated `README.md` to document the new `connections.yaml` structure, the `data_sources` section in the WBR YAML, and revised usage instructions.
- Adjusted `requirements.txt` to include necessary database client libraries and resolved a dependency conflict with `cryptography`.

The application no longer relies primarily on CSV uploads for metric data, offering a more robust and flexible way to integrate with existing data warehouses and databases. The `/download_yaml` feature now produces a template that requires users to configure the `data_sources` section manually for database integration.

google-labs-jules[bot] committed:
feat: Integrate database connectors for WBR data sourcing

This commit introduces the capability for the WBR application to fetch
metrics data directly from databases, supporting Amazon Athena, PostgreSQL,
Redshift, and Snowflake.

Key changes:
- Added a `connections.yaml` file to define database connection parameters.
- Modified the main WBR YAML configuration to include a `data_sources`
  section, where users specify a connection name (from `connections.yaml`),
  a SQL query, and the date column from the query result.
- Implemented database connector classes in `src/connectors/` for each
  supported database, along with a base connector and a factory function.
- Refactored `WBRValidator` to use these connectors to load data into a
  pandas DataFrame, replacing the previous CSV file input mechanism.
- Updated `controller.py` to handle the new configuration flow, loading
  `connections.yaml` at startup and passing data source definitions to
  the validator.
- Added unit tests for the new connectors, configuration loading, and
  related utility functions.
- Updated `README.md` to document the new `connections.yaml` structure,
  the `data_sources` section in the WBR YAML, and revised usage instructions.
- Adjusted `requirements.txt` to include necessary database client
  libraries and resolved a dependency conflict with `cryptography`.

The application no longer relies primarily on CSV uploads for metric data,
offering a more robust and flexible way to integrate with existing data
warehouses and databases. The `/download_yaml` feature now produces a
template that requires users to configure the `data_sources` section
manually for database integration.

ColinB-WB commented:
@OElesin, Thank you for this excellent pull request that adds database connectivity to the WBR App. This is a significant enhancement that will benefit users who want to integrate with existing data sources. The overall architecture and direction are promising and align well with our project's goals. We're currently reviewing the implementation details and will provide some feedback and suggestions shortly. We appreciate the thoroughness of your work, including the updated documentation and test coverage.

ColinB-WB commented:
@OElesin   
Thanks again for your thoughtful contribution to the WBR App and for introducing database-backed input support. After reviewing the changes, we’d like to request a few modifications to ensure the application remains robust, backward-compatible, and user-friendly for both technical and non-technical users.

Key Issues Identified:
1. Dependency on Local connections.yaml
We want to remove the requirement for users to manually place a connections.yaml file in the root directory.

2. Loss of CSV Support
We need to maintain backwards compatibility for users who use CSV files and do not have a database.

3. Current data_sources format
The current format, using a list structure, is harder to read and manage, especially when handling multiple queries and connections. We suggest moving to a cleaner, dictionary-based format (detailed below).


There are three changes, we are requesting.

1. db_config_url Support
The YAML config should include an optional field:

db_config_url: "https://example.com/path/to/connections.yaml"

This URL or file path points to a remotely hosted connections.yaml file with connection details.


2. YAML Input and Data Source Fallback Logic
The application will require a YAML config file as input and accept an optional CSV file.


The data loading behavior should follow this priority:


- CSV Present (Takes Precedence):
 If the user uploads a CSV file—even if db_config_url is configured in YAML—the app should use the CSV file as the data source. This allows users to override DB behavior easily.


- No CSV, but db_config_url Present:
 The app should fetch and parse the hosted connections.yaml, connect to the appropriate DB, and run the specified queries to retrieve data.


- Neither CSV nor db_config_url Present:
 The application should throw a clear exception (e.g., "No data source provided. Please provide either a CSV file or a db_config_url in your YAML config.")



3. Connection.yaml changes
To improve readability and scalability, please maket the following changes:

- Update the data_sources structure from list-based to dictionary-based format, 
- Remove connection_name,
- Add a description field to the query.
- Remove the date_column: line
- Change the query so the first column is always aliased as “Date”

Here’s the proposed structure:
data_sources:
  MyProdPostgres:
    queries:
      main_metrics:
        description: "Aggregates total sales and user count by day" 
        date_column: "event_date" # remove this line
        query: >
          SELECT
            Event_date as “Date”, #alias as “Date”
            SUM(sales_value) as total_sales,
            COUNT(DISTINCT user_id) as active_users
          FROM daily_transactions
          GROUP BY event_date

Let us know if you need clarification on any part. Once these changes are made, we’d be happy to move forward with merging the PR. Appreciate your work on improving the project!

google-labs-jules[bot] committed:
refactor: Implement DB connector and config feedback

This commit refactors the database connector feature based on pull request review feedback to improve usability, backward compatibility, and configuration management.

Key changes:
- **Re-introduced CSV support with fallback logic**: The application now prioritizes a direct CSV upload for data sourcing. If no CSV is provided, it falls back to using a database configuration specified in the WBR YAML. An error is raised if no data source is available.
- **Implemented on-demand DB config loading**: Removed the requirement for a local `connections.yaml` file. The WBR YAML config can now specify a `db_config_url` (a URL or file path) to a remotely hosted connections file, which is loaded on-demand.
- **Improved `data_sources` configuration**: The format in the WBR YAML was changed from a list to a more readable dictionary structure, where the keys are the connection names.
- **Enforced 'Date' column convention**: Removed the `date_column` field. The new convention requires that the first column of a data source SQL query must be the date column and must be aliased as `"Date"`. This is now validated in the base connector.
- **Updated Documentation**: The `README.md` has been rewritten to explain the new CSV-first fallback logic, the `db_config_url` field, the new `data_sources` format, and the "Date" column alias requirement.
- **Updated Unit Tests**: All unit tests for connectors, utilities, and the validator have been updated to reflect these changes, ensuring the new logic is covered.

OElesin commented:
refactor: Implement DB connector and config feedback

This commit refactors the database connector feature based on pull request review feedback to improve usability, backward compatibility, and configuration management.

Key changes:
- **Re-introduced CSV support with fallback logic**: The application now prioritizes a direct CSV upload for data sourcing. If no CSV is provided, it falls back to using a database configuration specified in the WBR YAML. An error is raised if no data source is available.
- **Implemented on-demand DB config loading**: Removed the requirement for a local `connections.yaml` file. The WBR YAML config can now specify a `db_config_url` (a URL or file path) to a remotely hosted connections file, which is loaded on-demand.
- **Improved `data_sources` configuration**: The format in the WBR YAML was changed from a list to a more readable dictionary structure, where the keys are the connection names.
- **Enforced 'Date' column convention**: Removed the `date_column` field. The new convention requires that the first column of a data source SQL query must be the date column and must be aliased as `"Date"`. This is now validated in the base connector.
- **Updated Documentation**: The `README.md` has been rewritten to explain the new CSV-first fallback logic, the `db_config_url` field, the new `data_sources` format, and the "Date" column alias requirement.
- **Updated Unit Tests**: All unit tests for connectors, utilities, and the validator have been updated to reflect these changes, ensuring the new logic is covered.

prasad700 commented on src/validator.py:
```diff
@@ -1,21 +1,114 @@
+            self.daily_df = pd.read_csv(csv_data, parse_dates=['Date'], thousands=',').sort_values(by='Date')
+        else:
+            logger.info("No CSV data provided. Attempting to load data from database source.")
+            db_config_url = self.cfg.get("db_config_url")
```
db_config_url should be retrieved from the setup section in the config file. Please update the code accordingly.

prasad700 commented on src/controller.py:
```diff
@@ -39,24 +39,30 @@ def get_wbr_metrics():
+        return app.response_class(response=json.dumps({"description": "configfile is required."}), status=400)
+
+    # CSV file is now optional
+    csv_data_file = request.files.get('csvfile')
```
A null check for the CSV file (data) should also be added in wbr.js at line 97, where the current code is:

formdata.append("csvfile", data, "WBR Sample Dataset.csv");

Currently, there is no validation to ensure that data is not null or undefined before appending it to the form. Please update the code to include a proper check to prevent potential runtime errors.

prasad-walvekar committed:
comment changes

prasad-walvekar committed:
test fixes

prasad-walvekar committed:
support for secret manager and fetch the credentials for vaults instead of expecting the credentials to be publicly open

prasad-walvekar committed:
removed unwanted parameter for DataLoader class

prasad-walvekar committed:
* Multiple datasource (Database and csv files) and multiple query handling
* Modified the AWS environment variables
* Modified README.md with new data_sources configuration

prasad-walvekar committed:
ModuleImportError fix

prasad-walvekar committed:
removed hashicorp reference in README.md

prasad-walvekar committed:
metric section changes updated in README.md

OElesin commented:
Wow!!

Great news and thanks for finalizing this and merging!!

Best regards,
Lekan

On Mon, Dec 22, 2025 at 9:01 AM devvedminds ***@***.***>
wrote:

> Merged #11 <https://github.com/working-backwards/wbr-app/pull/11> into
> master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/working-backwards/wbr-app/pull/11#event-21698956725>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWSYYRD2N5LCRHLWWPZH7L4C6QNJAVCNFSM6AAAAACAKUVNNWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMRRGY4TQOJVGY3TENI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Olalekan Fuad Elesin
***@***.***
(+​49​) 0​1744​ ​822 370​
<https://www.linkedin.com/in/elesinolalekan/>
<https://twitter.com/elesinOlalekan>
